packages/hardhat-core/src/builtin-tasks/flatten.ts

```diff
@@ -56,35 +61,64 @@ function getFileWithoutImports(resolvedFile: ResolvedFile) {
 
 subtask(
   TASK_FLATTEN_GET_FLATTENED_SOURCE,
-  "Returns all contracts and their dependencies flattened",
-  async (_, { run }) => {
+  "Returns all contracts and their dependencies flattened"
+)
+  .addOptionalParam("files", undefined, undefined, types.any)
+  .setAction(async ({ files }: { files?: string[] }, { run }) => {
+    const dependencyGraph: DependencyGraph = await run(
+      TASK_FLATTEN_GET_DEPENDENCY_GRAPH,
+      { files }
+    );
+
     let flattened = "";
 
-    const graph: DependencyGraph = await run(
-      TASK_COMPILE_SOLIDITY_GET_DEPENDENCY_GRAPH
-    );
-    if (graph.getResolvedFiles().length === 0) {
+    if (dependencyGraph.getResolvedFiles().length === 0) {
       return flattened;
     }
 
     const packageJson = await getPackageJson();
     flattened += `// Sources flattened with hardhat v${packageJson.version} https://usehardhat.com`;
 
-    const sortedFiles = getSortedFiles(graph);
+    const sortedFiles = getSortedFiles(dependencyGraph);
 
     for (const file of sortedFiles) {
       flattened += `\n\n// File ${file.getVersionedName()}\n`;
       flattened += `\n${getFileWithoutImports(file)}\n`;
     }
 
     return flattened.trim();
-  }
-);
+  });
+
+subtask(TASK_FLATTEN_GET_DEPENDENCY_GRAPH)
+  .addOptionalParam("files", undefined, undefined, types.any)
+  .setAction(async ({ files }: { files: string[] | undefined }, { run }) => {
+    const sourcePaths =
+      files === undefined
+        ? await run(TASK_COMPILE_SOLIDITY_GET_SOURCE_PATHS)
+        : files.map((f) => fs.realpathSync(f));
```

Comment:
Maybe we can accept globs here? Does that make sense?

Reply:
I'm curious about why you'd do that. Can't you just use the globs of the shell? Is it because that npm scripts limitation?

Anyway, people who always asked for this was to flatten a single contract actually.

Reply:
Not all shells support powerful globs (e.g. `contracts/legacy/**/*.sol`). And I think the "you can use either files or globs here" is a solid pattern, easy to implement and without any obvious downside (that I know at least).

That being said, it doesn't seem too important in this case, and it can be added later in a backward compatible way (unless someone has a file called `*.sol`, but in that case they deserve it).

Reply:
Fair point. Maybe we should add it in the future to all our tasks.